src/database.ts

```diff
@@ -1386,6 +1485,54 @@ class Database extends GrpcServiceObject {
     });
     return waitForSessionStream;
   }
+
+  restore(backupPath: string): Promise<RestoreDatabaseResponse>;
+  restore(backupPath: string, callback: RestoreDatabaseCallback): void;
+  /**
+   * Restore a backup into this database.
+   *
+   * When this call completes, the restore will have commenced but will not
+   * necessarily have completed.
+   *
+   * @param backupPath the path of the backup to restore.
+   * @returns Promise<RestoreDatabaseResponse> when resolved, contains the restore operation.
+   *
+   * @example
+   * const {Spanner} = require('@google-cloud/spanner');
+   * const spanner = new Spanner();
+   * const instance = spanner.instance('my-instance');
+   * const database = instance.database('my-database');
+   * const [, restoreOperation] = await database.restore('projects/my-project/instances/my-instance/backups/my-backup');
+   * // Wait for restore to complete
+   * await restoreOperation.promise();
+   */
+  restore(
+    backupPath: string,
+    callback?: RestoreDatabaseCallback
+  ): Promise<RestoreDatabaseResponse> | void {
+    const reqOpts: databaseAdmin.spanner.admin.database.v1.IRestoreDatabaseRequest = extend(
+      {
+        parent: this.instance.formattedName_,
+        databaseId: this.id,
+        backup: backupPath,
+      }
+    );
+    return this.request(
+      {
+        client: 'DatabaseAdminClient',
+        method: 'restoreDatabase',
+        reqOpts,
+      },
+      (err, resp) => {
+        if (err) {
+          callback!(err, null, resp!);
+          return;
+        }
+        callback!(null, this, resp, resp);
```

Comment:
This is mis-matched from the error handler callback signature. Should we double up the `resp` there, or lose one here?

Reply:
The `LongRunningCallback<Database>` type needs four parameters in the callback. So I updated both to call it with four args.